mono/mini/method-to-ir.c

```diff
@@ -7036,8 +7006,22 @@ is_supported_tail_call (MonoCompile *cfg, MonoMethod *method, MonoMethod *cmetho
 		supported_tail_call = FALSE;
 	if (cmethod->wrapper_type && cmethod->wrapper_type != MONO_WRAPPER_DYNAMIC_METHOD)
 		supported_tail_call = FALSE;
-	if (call_opcode != CEE_CALL)
+
+	g_assert (call_opcode == CEE_CALL || call_opcode == CEE_CALLVIRT || call_opcode == CEE_CALLI);
+	switch (call_opcode) {
+	case CEE_CALL:
+		break;
+	case CEE_CALLI:
+		supported_tail_call = FALSE;
+		break;
+	case CEE_CALLVIRT: // FIXME finish all architectures, AOT, LLVM
+		if (!cfg->backend->have_op_tail_call_membase || cfg->compile_aot || cfg->llvm_only || COMPILE_LLVM (cfg))
+			supported_tail_call = FALSE;
+		break;
+	default:
 		supported_tail_call = FALSE;
```

Comment:
How can you even reach the `default` case here with the assertion at https://github.com/mono/mono/pull/7426/files#diff-e7526071878d708680e4a8c5fa5a46edR7010 ? It would be better in this case to remove the assertion and use a `g_error` in the default case with the value of `call_opcode` so it gives us some information in case of crash in the wild.